llvm/include/llvm/CAS/ActionCache.h

```diff
@@ -0,0 +1,100 @@
+//===- llvm/CAS/ActionCache.h -----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_CASACTIONCACHE_H
+#define LLVM_CAS_CASACTIONCACHE_H
+
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/CAS/CASReference.h"
+#include "llvm/Support/Error.h"
+
+namespace llvm::cas {
+
+class CASDB;
+class CASID;
+class ObjectProxy;
+
+/// A key for caching an operation.
+/// It is implemented as a bag of bytes and provided convenient constructor
+/// for string/CAS types.
+// TODO: Support CacheKey other than a CASID but rather any array of bytes.
+// To do that, ActionCache need to be able to rehash the key into the index,
+// which then `getOrCompute` method can be used to avoid multiple calls to
+// has function.
+class CacheKey {
+public:
+  StringRef getKey() const { return Key; }
+
+  CacheKey(const CASID &ID);
+  CacheKey(const ObjectProxy &Proxy);
+  CacheKey(const CASDB &CAS, const ObjectRef &Ref);
+
+private:
+  std::string Key;
+};
+
+/// A cache from a key describing an action to the result of doing it.
+///
+/// Actions are expected to be pure (write-once, collision is an error).
```

Comment:
It's only an error if the value changes, right? Maybe call that out since "write-once" makes it sound like `put`ing the same key-value pair twice could be an error.